probe-rs/src/flash/error.rs

```diff
@@ -0,0 +1,50 @@
+use thiserror::Error;
+
+use crate::config::FlashRegion;
+use crate::error;
+
+#[derive(Error, Debug)]
+pub enum FlashError {
+    #[error("The execution of '{name}' failed with code {errorcode}")]
+    RoutineCallFailed { name: &'static str, errorcode: u32 },
+    #[error("'{0}' is not supported")]
+    NotSupported(&'static str),
+    #[error("Buffer {n}/{max} does not exist")]
+    InvalidBufferNumber { n: usize, max: usize },
+    #[error("Something during memory interaction went wrong: {0}")]
+    Memory(#[source] error::Error),
+    #[error("Something during the interaction with the core went wrong: {0}")]
+    Core(#[source] error::Error),
+    #[error("{address} is not contained in {region:?}")]
+    AddressNotInRegion { address: u32, region: FlashRegion },
+    #[error(
+        "The RAM contents did not match the expected contents after loading the flash algorithm."
+    )]
+    FlashAlgorithmNotLoaded,
+    #[error(
+        "The page write of the page at address {page_address:#08X} failed with error code {error_code}."
```

Comment:
What does error code mean here? Could we change this to an enum?

Reply:
I guess the same applies here.

Reply:
Do we need to have different errors? Maybe we could merge them to a combined error, indication that a function of the flash algorithm failed.

Reply:
To be clear, I mean merging with `RoutineCallFailed`.